### PR TITLE
More on database consistency

### PR DESCRIPTION
**Pull request by jjos2372**
Smart contracts (ATs) can also *burn* burst and this needs to be accounted for in the DB consistency test.

Also includes other minor improvements and better logic to avoid DB trimming if any inconsistency is detected.

image

### CODE REVIEW

File: src/brs/BlockchainProcessorImpl.java [hunk 1]
```diff
@@ -149,12 +149,19 @@ public BlockchainProcessorImpl(ThreadPool threadPool, BlockService blockService,
     }, Event.BLOCK_PUSHED);
 
     blockListeners.addListener(block -> transactionProcessor.revalidateUnconfirmedTransactions(), Event.BLOCK_PUSHED);
-
+    if (trimDerivedTables) {    
+      blockListeners.addListener(block -> { 
+        if (block.getHeight() % Constants.MAX_ROLLBACK == 0) {
+          if (lastTrimHeight.get() > 0) {
```
Comment (ohager) [inline src/brs/BlockchainProcessorImpl.java:155]: I would join both conditions

File: src/brs/BlockchainProcessorImpl.java [hunk 2]
```diff
@@ -149,12 +149,19 @@ public BlockchainProcessorImpl(ThreadPool threadPool, BlockService blockService,
...
+          } 
+        }   
+      }, Event.AFTER_BLOCK_APPLY);  
+    }
     addGenesisBlock();
     if(Boolean.FALSE.equals(propertyService.getBoolean(Props.DB_SKIP_CHECK))) {
       if(checkDatabaseState() != 0) {
-        logger.error("Database is inconsistent, try to pop off 1000 blocks or more or sync from empty. Alternatively add '{} = true' at your own risk.", Props.DB_SKIP_CHECK.getName());
-        System.exit(-1);
+        logger.warn("Database is inconsistent, try to pop off to block height {} or sync from empty.", getMinRollbackHeight());
```
Comment (ohager) [inline src/brs/BlockchainProcessorImpl.java:164]: same here, but ofc it's a matter of preference

**Review by ohager — APPROVED** [review]
LGTM - but left a comment on nested ifs

File: src/brs/BlockchainProcessorImpl.java (hunk 2 above)
Comment (jjos2372) [reply src/brs/BlockchainProcessorImpl.java]: Agreed, will be adjusted and merged.

**Pull request merged**